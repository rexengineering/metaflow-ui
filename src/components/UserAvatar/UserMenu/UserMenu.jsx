import {
  Badge,
  makeStyles,
  Menu,
  MenuItem,
  Typography,
} from "@material-ui/core";
import PropTypes from "prop-types";
import React from "react";

const useStyles = makeStyles((theme) => ({
  underlinedText: {
    textDecoration: "underline",
  },
  listItem: {
    padding: theme.spacing(0.625, 2, 0),
  },
  menu: {
    marginTop: theme.spacing(4),
  },
  typographyToCapitalize: {
    textTransform: "capitalize",
  },
}));

function UserMenu({
  status,
  statuses,
  menuAnchor,
  onMenuClose,
  onStatusSelection,
}) {
  const classes = useStyles();
  const userStatusOptions = Object.keys(statuses);
  return (
    <Menu
      className={classes.menu}
      open={!!menuAnchor}
      anchorEl={menuAnchor}
      onClose={onMenuClose}
    >
      <MenuItem disabled className={classes.underlinedText}>
        Set your status
      </MenuItem>
      {Array.isArray(userStatusOptions) &&
        userStatusOptions.map((currentStatusKey) => {
          const isCurrentState = status === currentStatusKey;
          const typographyVariant = isCurrentState ? "body1" : "h6";
          return (
            <MenuItem
              onClick={() => onStatusSelection(currentStatusKey)}
              className={classes.listItem}
              key={currentStatusKey}
            >
              <Badge
                background={statuses[status]}
                variant="dot"
                invisible={isCurrentState}
              >
                <Typography
                  className={classes.typographyToCapitalize}
                  variant={typographyVariant}
                >
                  {currentStatusKey}
                </Typography>
              </Badge>
            </MenuItem>
          );
        })}
      <MenuItem disabled className={classes.underlinedText}>
        Account
      </MenuItem>
      <MenuItem>Profile</MenuItem>
      <MenuItem>Logout</MenuItem>
    </Menu>
  );
}

UserMenu.defaultProps = {
  menuAnchor: null,
};

UserMenu.propTypes = {
  status: PropTypes.string.isRequired,
  statuses: PropTypes.objectOf(PropTypes.any).isRequired,
  menuAnchor: PropTypes.objectOf(PropTypes.any),
  onMenuClose: PropTypes.func.isRequired,
  onStatusSelection: PropTypes.func.isRequired,
};

export default UserMenu;
