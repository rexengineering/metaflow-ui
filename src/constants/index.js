export const BUSY = "BUSY";
export const OFFLINE = "OFFLINE";
export const ONLINE = "ONLINE";

export const ActionItemDeleteType = "ActionDeleteType";
export const ActionItemCheckType = "ActionCheckType";
export const ActionItemDisplayOnlyType = "ActionDisplayOnlyType";

export const InitialButtonState = 'initial';
export const LoadingButtonState = 'loading';
export const talkTrackIdentifierProp = 'iid';
export const talkTrackType = 'talktrack';