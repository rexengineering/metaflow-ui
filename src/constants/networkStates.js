export const INITIAL = "INITIAL";
export const REQUEST = "REQUEST";
export const SUCCESS = "SUCCESS";
export const FAILURE = "FAILURE";

export default {
  INITIAL,
  REQUEST,
  SUCCESS,
  FAILURE
}
