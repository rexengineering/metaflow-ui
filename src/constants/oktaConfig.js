const config = {
  oidc: {
    clientId: window.REX.OKTA.CLIENT_ID,
    issuer: window.REX.OKTA.ISSUER,
    redirectUri: window.REX.OKTA.REDIRECT_URI,
    scopes: ["openid", "profile", "email"],
    pkce: true,
    disableHttpsCheck: window.REX.OKTA.OKTA_TESTING_DISABLEHTTPSCHECK,
  },
  resourceServer: {
    messagesUrl: "http://localhost:5000/api/messages",
  },
};

export default config;
