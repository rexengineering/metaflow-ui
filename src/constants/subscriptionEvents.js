export const START_WORKFLOW = "START_WORKFLOW";
export const UPDATE_WORKFLOW = "UPDATE_WORKFLOW";
export const FINISH_WORKFLOW = "FINISH_WORKFLOW";
export const START_TASK = "START_TASK";
export const UPDATE_TASK = "UPDATE_TASK";
export const FINISH_TASK = "FINISH_TASK";
export const START_BROADCAST = "START_BROADCAST";
export const FINISH_BROADCAST = "FINISH_BROADCAST";
export const ERROR_BROADCAST = "ERROR_BROADCAST";
export const KEEP_ALIVE_SUCCESS = "SUCCESS";
export const KEEP_ALIVE_FAILURE = "FAILURE";