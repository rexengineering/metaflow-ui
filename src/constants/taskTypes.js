import { string, number, boolean } from "yup";

// Inputs
export const TEXT = "TEXT";
export const PHONE_NUMBER = "PHONE_NUMBER";
export const PERCENTAGE = "PERCENTAGE";
export const CURRENCY = "CURRENCY";
export const BOOLEAN = "BOOLEAN";
export const INTEGER = "INTEGER";
export const FLOAT = "FLOAT";

export function isInputType(field) {
  return [
    TEXT,
    PHONE_NUMBER,
    PERCENTAGE,
    CURRENCY,
    BOOLEAN,
    INTEGER,
    FLOAT,
  ].includes(field);
}

// Info
export const TABLE = "TABLE";
export const COPY = "COPY";
export const WORKFLOW_INSTANTIATOR = "WORKFLOW";

export function isInfoType(type) {
  return [TABLE, COPY, WORKFLOW_INSTANTIATOR].includes(type);
}

const numberValidationMessage = "Please enter a number";

export const validationSchemaMapping = {
  TEXT: string("Please enter a value"),
  PHONE_NUMBER: string("Please enter a value").length(10),
  PERCENTAGE: number(numberValidationMessage).typeError(
    numberValidationMessage
  ),
  CURRENCY: number(numberValidationMessage).typeError(numberValidationMessage),
  BOOLEAN: boolean(),
  INTEGER: number(numberValidationMessage)
    .typeError(numberValidationMessage)
    .integer(),
  FLOAT: number(numberValidationMessage).typeError(numberValidationMessage),
};

const taskTypes = {
  TEXT,
  PHONE_NUMBER,
  PERCENTAGE,
  CURRENCY,
  BOOLEAN,
  INTEGER,
  FLOAT,
  TABLE,
  COPY,
  WORKFLOW_INSTANTIATOR,
};

export default taskTypes;
