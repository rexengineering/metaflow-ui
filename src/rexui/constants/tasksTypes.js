import { string, number, boolean } from "yup";


const numberValidationMessage = "Please enter a number";

export const TEXT = "TEXT";
export const PHONE_NUMBER = "PHONE_NUMBER";
export const PERCENTAGE = "PERCENTAGE";
export const CURRENCY = "CURRENCY";
export const BOOLEAN = "BOOLEAN";
export const INTEGER = "INTEGER";
export const FLOAT = "FLOAT";
export const TABLE = "TABLE";
export const COPY = "COPY";
export const WORKFLOW_INSTANTIATOR = "WORKFLOW";

export const validationSchemaMapping = {
    TEXT: string("Please enter a value"),
    PHONE_NUMBER: string("Please enter a value").length(10),
    PERCENTAGE: number(numberValidationMessage).typeError(numberValidationMessage),
    CURRENCY: number(numberValidationMessage).typeError(numberValidationMessage),
    BOOLEAN: boolean(),
    INTEGER: number(numberValidationMessage)
        .typeError(numberValidationMessage)
        .integer(),
    FLOAT: number(numberValidationMessage).typeError(numberValidationMessage),
};

export function isInfoType(type) {
    return [TABLE, COPY, WORKFLOW_INSTANTIATOR].includes(type);
}

export function isInputType(field) {
    return [
        TEXT,
        PHONE_NUMBER,
        PERCENTAGE,
        CURRENCY,
        BOOLEAN,
        INTEGER,
        FLOAT,
    ].includes(field);
}