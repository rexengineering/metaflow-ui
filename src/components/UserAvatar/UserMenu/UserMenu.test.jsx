import { fireEvent, render, screen, waitFor } from "@testing-library/react";
import React from "react";
import { Typography } from "@material-ui/core";
import { BUSY, OFFLINE, ONLINE } from "../../../constants";
import UserMenu from "./UserMenu";

describe("<UserMenu />", () => {
  const mockedStatuses = {
    [ONLINE]: "ONLINE",
    [OFFLINE]: "OFFLINE",
    [BUSY]: "BUSY",
  };

  const props = {
    status: BUSY,
    statuses: mockedStatuses,
    menuAnchor: <Typography>Menu</Typography>,
    onMenuClose: () => {},
    onStatusSelection: () => {},
  };

  const renderUserMenu = ({
    status,
    statuses,
    menuAnchor,
    onMenuClose,
    onStatusSelection,
  } = props) => (
    <UserMenu
      status={status}
      statuses={statuses}
      menuAnchor={menuAnchor}
      onMenuClose={onMenuClose}
      onStatusSelection={onStatusSelection}
    />
  );

  it("should render", () => {
    const { asFragment } = render(renderUserMenu());
    const initialRender = asFragment();

    expect(initialRender).not.toBeNull();
  });

  it("should match snapshot", () => {
    const { asFragment } = render(renderUserMenu());
    const initialRender = asFragment();

    expect(initialRender).toMatchSnapshot();
  });

  it("should render all given statuses", () => {
    render(renderUserMenu());
    const statusesKeys = Object.keys(mockedStatuses);
    statusesKeys.forEach((status) =>
      expect(screen.getByText(status)).toBeInTheDocument()
    );
  });

  it("should handle onClick properly", async () => {
    const onStatusSelection = jest.fn();
    render(renderUserMenu({ ...props, onStatusSelection }));
    const statusesKeys = Object.keys(mockedStatuses);

    await waitFor(() =>
      statusesKeys.forEach((status) => {
        const statusNode = screen.getByText(status);
        fireEvent.click(statusNode);
      })
    );
    expect(onStatusSelection).toHaveBeenCalledTimes(statusesKeys.length);
  });
});
