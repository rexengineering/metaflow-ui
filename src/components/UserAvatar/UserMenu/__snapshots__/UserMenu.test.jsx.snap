// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<UserMenu /> should match snapshot 1`] = `<DocumentFragment />`;
